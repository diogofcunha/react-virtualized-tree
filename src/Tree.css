.tree-node {
  user-select: none;
  cursor: pointer;
}
