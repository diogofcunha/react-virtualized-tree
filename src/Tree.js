import React from 'react';
import PropTypes from 'prop-types';
import { AutoSizer, List } from 'react-virtualized';

import { FlattenedNode } from './shapes/nodeShapes';

export default class Tree extends React.Component {
  rowRenderer =  (nodes) => ({ key, index, style}) => {
    const { children: NodeRenderer } = this.props;
    const node = nodes[index];
    return (
      <div
        key={key}
        className="tree-node"
        style={{ ...style, marginLeft: node.deepness * 30}}
      >
        <NodeRenderer node={node} onChange={this.props.onChange}/>
      </div>
    )
  }

  render() {
    const { children, nodes } = this.props;

    return (
      <AutoSizer>
        {({ height, width }) => (
        <List
            height={height}
            rowCount={nodes.length}
            rowHeight={21}
            rowRenderer={this.rowRenderer(nodes)}
            width={width}
        />
        )}
      </AutoSizer>
    );
  }
}

Tree.propTypes = {
  nodes: PropTypes.arrayOf(PropTypes.shape(FlattenedNode)).isRequired,
  children: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
};
