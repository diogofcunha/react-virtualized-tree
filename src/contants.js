export const UPDATE_TYPE = {
  ADD: 0,
  DELETE: 1,
  UPDATE: 2,
  MOVE_AS_CHILDREN: 3,
  MOVE_AS_SIBLING: 4
}