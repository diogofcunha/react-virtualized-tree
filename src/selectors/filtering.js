const INITIAL_FILTERED_VALUE = {nodes: [], nodeParentMappings: {}};

export const filterNodes = (filter, nodes, parents = []) =>
  nodes.reduce((filtered, n) => {
    const {nodes: filteredChildren, nodeParentMappings: childrenNodeMappings} = n.children
      ? filterNodes(filter, n.children, [...parents, n.id])
      : INITIAL_FILTERED_VALUE;

    return !(filter(n) || filteredChildren.length)
      ? filtered
      : {
          nodes: [
            ...filtered.nodes,
            {
              ...n,
              children: filteredChildren,
            },
          ],
          nodeParentMappings: {
            ...filtered.nodeParentMappings,
            ...childrenNodeMappings,
            [n.id]: parents,
          },
        };
  }, INITIAL_FILTERED_VALUE);
