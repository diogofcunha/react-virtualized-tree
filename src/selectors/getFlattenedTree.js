export const isNodeExpanded = node => node.state && node.state.expanded;
export const nodeHasChildren = node => node.children && node.children.length;

export const getFlattenedTree = (nodes, parents = []) =>
  nodes.reduce((flattenedTree, node) => {
    const deepness = parents.length;
    const nodeWithHelpers = {...node, deepness, parents};

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      return [...flattenedTree, nodeWithHelpers];
    }

    return [...flattenedTree, nodeWithHelpers, ...getFlattenedTree(node.children, [...parents, node.id])];
  }, []);

export const getFlattenedTreePaths = (nodes, parents = []) => {
  const paths = [];

  for (const node of nodes) {
    const {id} = node;

    if (!nodeHasChildren(node) || !isNodeExpanded(node)) {
      paths.push(parents.concat(id));
    } else {
      paths.push(parents.concat(id));
      paths.push(...getFlattenedTreePaths(node.children, [...parents, id]));
    }
  }

  return paths;
};

export const doesChangeAffectFlattenedTree = (previousNode, nextNode) => {
  return isNodeExpanded(previousNode) !== isNodeExpanded(nextNode);
};
