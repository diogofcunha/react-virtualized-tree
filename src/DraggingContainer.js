import React from 'react';
import PropTypes from 'prop-types';

import { Node } from './shapes/nodeShapes';
import { UPDATE_TYPE } from './contants';

export default class DraggingContainer extends React.Component {
  state = {
    fromNode: {}
  }

  handleDrag = fromNode => {
    this.setState({ fromNode });
  }

  handleDrop = (toNode, type) => {
    this.props.onChange({
      type,
      node: [ this.state.fromNode, toNode ]
    });

    this.setState({ fromNode: {} });
  }

  getChildContext = () => ({
    handleDrag: this.handleDrag,
    handleDrop: this.handleDrop,
    fromNode: this.state.fromNode
  })

  static childContextTypes = {
    handleDrag: PropTypes.func,
    handleDrop: PropTypes.func,
    fromNode: PropTypes.object
  }

  render() {
    return (
      <div
        className={this.props.className}
        onDrop={() => this.handleDrop({ ...this.state.fromNode, parents: [] }, UPDATE_TYPE.MOVE_AS_SIBLING)}
        onDragOver={e => e.preventDefault()}
      > 
        { this.props.children }
      </div>
    )
  }
}

DraggingContainer.propTypes = {
  onChange: PropTypes.func.isRequired,
  className: PropTypes.string.isRequired,
  children: PropTypes.element.isRequired
}