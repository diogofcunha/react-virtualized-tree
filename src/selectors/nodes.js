import { createSelector } from 'reselect';
import { omit } from 'lodash';

import { UPDATE_TYPE } from '../contants';

export const getNodeRenderOptions = createSelector(
  (node => (node.state || {}).expanded),
  (node => (node.state || {}).favorite),
  (node => (node.state || {}).deletable),
  (node => node.children), 
  (expanded, favorite, deletable, children = []) => ({
    hasChildren: !!children.length,
    isExpanded: !!expanded,
    isFavorite: !!favorite,
    isDeletable: !!deletable
  })
);

const FLATTEN_TREE_PROPERTIES = [ 'deepness', 'parents' ];
const NODE_OPERATION_PROPERTIES = [ 'index' ];

const NODE_OPERATION_TYPES = {
  CHANGE_NODE: 'CHANGE_NODE',
  DELETE_NODE: 'DELETE_NODE',
  ADD_NODE: 'ADD_NODE'
}

const NODE_CHANGE_OPERATIONS = {
  CHANGE_NODE: (nodes, updatedNode) => nodes.map(n => n.id === updatedNode.id ? omit(updatedNode, FLATTEN_TREE_PROPERTIES) : n),
  DELETE_NODE: (nodes, updatedNode) => nodes.filter(n => n.id !== updatedNode.id),
  ADD_NODE: (nodes, nodeToAdd) => [ ...nodes, omit(nodeToAdd, [ ...FLATTEN_TREE_PROPERTIES, ...NODE_OPERATION_PROPERTIES] ) ]
}

export const replaceNodeFromTree = (nodes, updatedNode, operation = NODE_OPERATION_TYPES.CHANGE_NODE) => {
  if (!NODE_CHANGE_OPERATIONS[operation]) {
    return nodes;
  }
  
  const { parents } = updatedNode;

  if (!parents.length) {
    return NODE_CHANGE_OPERATIONS[operation](nodes, updatedNode);
  }

  const parentIndex = nodes.findIndex(n => n.id === parents[0])

  const preSiblings = nodes.slice(0, parentIndex);
  const postSiblings = nodes.slice(parentIndex + 1);

  return [
    ...preSiblings,
    {
      ...nodes[parentIndex], 
      ...nodes[parentIndex].children || operation === NODE_OPERATION_TYPES.ADD_NODE ? 
        { children: replaceNodeFromTree(
            nodes[parentIndex].children || [],
            { ...updatedNode, parents: parents.slice(1) },
            operation
          )
        } : {}
    },
    ...postSiblings
  ]
}

export const deleteNodeFromTree = (nodes, deletedNode) => {
  return replaceNodeFromTree(
    nodes,
    deletedNode,
    NODE_OPERATION_TYPES.DELETE_NODE
  );
}

export const addNodeToTree = (nodes, nodeToAdd) => replaceNodeFromTree(
  nodes,
  nodeToAdd,
  NODE_OPERATION_TYPES.ADD_NODE
)

export const moveNodeToChildren = (nodes, [ fromNode, toNode ]) => {
  const { id: targetId, parents: targetParents } = toNode;
  const parents = [ ...targetParents, ...targetId !== fromNode.id ? [targetId] : []];

  return moveNodeFromTree(nodes, [ fromNode, { ...toNode, parents } ]);
}

export const moveNodeFromTree = (nodes, [ fromNode, toNode ]) => {
  const treeWithoutMovedNode = deleteNodeFromTree(
    nodes,
    fromNode
  );

  return addNodeToTree(treeWithoutMovedNode, { ...fromNode, parents: toNode.parents });
}

export const udpateNode = (originalNode, newState) => ({
  node: {
    ...originalNode,
    state : {
      ...originalNode.state,
      ...newState
    }
  },
  type: UPDATE_TYPE.UPDATE
});

export const deleteNode = node => ({
  node,
  type: UPDATE_TYPE.DELETE
});

export const addNode = node => ({
  node,
  type: UPDATE_TYPE.ADD
});
