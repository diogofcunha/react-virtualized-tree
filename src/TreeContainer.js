import React from 'react';
import PropTypes from 'prop-types';

import Tree from './Tree';
import DraggingContainer from './DraggingContainer';
import { UPDATE_TYPE } from './contants';
import { getFlattenedTree } from './selectors/getFlattenedTree';
import { deleteNodeFromTree, replaceNodeFromTree, moveNodeFromTree, moveNodeToChildren } from './selectors/nodes';
import { Node } from './shapes/nodeShapes';

const DEFAULT_UPDATE_TYPES = {
  [UPDATE_TYPE.DELETE]: deleteNodeFromTree,
  [UPDATE_TYPE.UPDATE]: replaceNodeFromTree,
  [UPDATE_TYPE.MOVE_AS_SIBLING]: moveNodeFromTree,
  [UPDATE_TYPE.MOVE_AS_CHILDREN]: moveNodeToChildren
};

export default class TreeContainer extends React.Component {
  constructor(props) {
    super(props);

    this.extensions = props;
  }

  static contextTypes = {
    unfilteredNodes: PropTypes.arrayOf(PropTypes.shape(Node))
  };

  get extensions() {
    return this._extensions;
  }

  set extensions(props) {
    const {
      extensions: {
        updateTypeHandlers = { }
      } = { } 
    } = props;

    this._extensions = {
      updateTypeHandlers: {
        ...DEFAULT_UPDATE_TYPES,
        ...updateTypeHandlers
      }
    }
  }

  get nodes() {
    return this.context.unfilteredNodes || this.props.nodes;
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.extensions !== this.props.extensions) {
      this.extensions = nextProps;
    }
  }

  handleChange = ({ node, type }) => {
    const updatedNodes = this.extensions.updateTypeHandlers[type](this.nodes, node);

    this.props.onChange(updatedNodes);
  }

  getContainer = () => !this.props.draggable ?
    { Container: 'div', props: {} } : 
    { Container: DraggingContainer, props: { onChange: this.handleChange } }

  render() {
    const { children, nodes } = this.props;
    const { Container, props } = this.getContainer();

    return (
      <Container {...props} className="tree-container" >
        <Tree nodes={getFlattenedTree(nodes)} onChange={this.handleChange}> 
          { children }
        </Tree>
      </Container>
    );
  }
};

TreeContainer.propTypes = {
  extensions: PropTypes.shape({
    updateTypeHandlers: PropTypes.object
  }),
  nodes: PropTypes.arrayOf(PropTypes.shape(Node)).isRequired,
  onChange: PropTypes.func,
  draggable: PropTypes.bool,
  children: PropTypes.func.isRequired
};
