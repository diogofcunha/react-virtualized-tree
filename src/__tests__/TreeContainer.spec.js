import React from 'react';
import { shallow } from 'enzyme';
import { omit } from 'lodash';

import TreeContainer from '../TreeContainer';
import DraggingContainer from '../DraggingContainer';
import Tree from '../Tree';
import { Nodes } from '../../testData/sampleTree';
import { getFlattenedTree } from '../selectors/getFlattenedTree';
import { UPDATE_TYPE } from '../contants';
import { replaceNodeFromTree, udpateNode } from '../selectors/nodes';

describe('TreeContainer', () => {
  const setup = (children = () => <span></span>, extraProps = {}, context = {}) => {
    const defaultProps = {
      nodes: Nodes,
      onChange: jest.fn()
    };

    const props = { ...defaultProps, ...extraProps };

    const wrapper = shallow(
      <TreeContainer {...props}>
        { children }
      </TreeContainer>,
      { context }
    );

    return {
      wrapper,
      props,
      treeWrapper: wrapper.find(Tree)
    }
  };

  it('should render a Tree with children and the correct props', () => {
    const exampleChild = jest.fn();
    const { treeWrapper, wrapper } = setup(exampleChild);
    const { nodes } = treeWrapper.props();

    expect(nodes).toMatchSnapshot();
  });

  it('when draggable should render a DraggingContainer', () => {
    const exampleChild = jest.fn();
    const { treeWrapper, wrapper } = setup(exampleChild, { draggable: true });

    expect(
      omit(wrapper.find(DraggingContainer).props(), 'children')
    ).toMatchSnapshot();
  });

  describe('change handle', () => {
    const getSampleNode = () => getFlattenedTree(Nodes)[2]

    it('should call injected prop onChange with the correct params when a node is deleted', () => {
      const { treeWrapper, props } = setup();
      treeWrapper.simulate('change', { node: getSampleNode(), type: UPDATE_TYPE.DELETE });

      expect(
        props.onChange.mock.calls[0]
      ).toMatchSnapshot();
    });

    it('should call injected prop onChange with the correct params when a node is updated', () => {
      const { treeWrapper, props } = setup();

      treeWrapper.simulate('change', {
        node: { ...getSampleNode(), state: { favorite: false, deletable: false } },
        type: UPDATE_TYPE.UPDATE 
      });

      expect(
        props.onChange.mock.calls[0]
      ).toMatchSnapshot();
    });

    it('should call injected prop onChange with the correct params when a node is updated when nodes come from context', () => {
      const { treeWrapper, props } = setup(
        jest.fn(),
        { nodes: Nodes.filter((n, i) => i % 2 === 0) },
        { unfilteredNodes: Nodes }
      );

      treeWrapper.simulate('change', {
        node: { ...getSampleNode(), state: { favorite: false, deletable: false } },
        type: UPDATE_TYPE.UPDATE 
      });

      expect(
        props.onChange.mock.calls[0]
      ).toMatchSnapshot();
    });
  });

  describe('extensions', () => {
    const getSampleNode = () => getFlattenedTree(Nodes)[2];

    describe('updateTypeHandlers', () => {
      it('should call injected prop onChange with the correct params for custom handlers', () => {
        const EXPAND_ALL = 3;

        const expandAll = nodes => nodes.map(node => ({
          ...node,
          state: {
            ...node.state,
            expanded: true
          },
          children: node.children ? expandAll(node.children) : []
        }));

        const { treeWrapper, props } = setup(() => <span></span>, { 
          extensions: {
            updateTypeHandlers: {
              [EXPAND_ALL]: expandAll
            }
          }
        });
  
        treeWrapper.simulate('change', { node: getSampleNode(), type: EXPAND_ALL });
  
        expect(
          props.onChange.mock.calls[0]
        ).toMatchSnapshot();
      });

      it('should call injected prop onChange with the correct params for handler overrides', () => {
        const updateAndFlag = (nodes, updatedNode) => replaceNodeFromTree(
          nodes,
          { ...updatedNode, updated: true }
        );

        const { treeWrapper, props } = setup(() => <span></span>, { 
          extensions: {
            updateTypeHandlers: {
              [UPDATE_TYPE.UPDATE]: updateAndFlag
            }
          }
        });
  
        treeWrapper.simulate('change', { node: getSampleNode(), type: UPDATE_TYPE.UPDATE });
  
        expect(
          props.onChange.mock.calls[0]
        ).toMatchSnapshot();
      });
    });
  });
});
