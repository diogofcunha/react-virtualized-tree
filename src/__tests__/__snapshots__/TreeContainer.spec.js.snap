// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeContainer change handle should call injected prop onChange with the correct params when a node is deleted 1`] = `
Array [
  Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": 4,
              "name": "Leaf 4",
            },
          ],
          "id": 2,
          "name": "Leaf 2",
          "state": Object {
            "deletable": true,
            "expanded": true,
          },
        },
        Object {
          "id": 5,
          "name": "Leaf 5",
        },
      ],
      "id": 0,
      "name": "Leaf 1",
      "state": Object {
        "expanded": true,
      },
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": 7,
              "name": "Leaf 8",
            },
            Object {
              "id": 8,
              "name": "Leaf 9",
            },
          ],
          "id": 6,
          "name": "Leaf 7",
          "state": Object {
            "expanded": false,
          },
        },
        Object {
          "id": 9,
          "name": "Leaf 10",
        },
      ],
      "id": 1,
      "name": "Leaf 6",
      "state": Object {
        "deletable": true,
        "expanded": false,
      },
    },
    Object {
      "id": "z",
      "name": "Leaf z",
      "state": Object {
        "deletable": true,
        "favorite": true,
      },
    },
  ],
]
`;

exports[`TreeContainer change handle should call injected prop onChange with the correct params when a node is updated 1`] = `
Array [
  Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": 3,
              "name": "Leaf 3",
              "state": Object {
                "deletable": false,
                "favorite": false,
              },
            },
            Object {
              "id": 4,
              "name": "Leaf 4",
            },
          ],
          "id": 2,
          "name": "Leaf 2",
          "state": Object {
            "deletable": true,
            "expanded": true,
          },
        },
        Object {
          "id": 5,
          "name": "Leaf 5",
        },
      ],
      "id": 0,
      "name": "Leaf 1",
      "state": Object {
        "expanded": true,
      },
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": 7,
              "name": "Leaf 8",
            },
            Object {
              "id": 8,
              "name": "Leaf 9",
            },
          ],
          "id": 6,
          "name": "Leaf 7",
          "state": Object {
            "expanded": false,
          },
        },
        Object {
          "id": 9,
          "name": "Leaf 10",
        },
      ],
      "id": 1,
      "name": "Leaf 6",
      "state": Object {
        "deletable": true,
        "expanded": false,
      },
    },
    Object {
      "id": "z",
      "name": "Leaf z",
      "state": Object {
        "deletable": true,
        "favorite": true,
      },
    },
  ],
]
`;

exports[`TreeContainer change handle should call injected prop onChange with the correct params when a node is updated when nodes come from context 1`] = `
Array [
  Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": 3,
              "name": "Leaf 3",
              "state": Object {
                "deletable": false,
                "favorite": false,
              },
            },
            Object {
              "id": 4,
              "name": "Leaf 4",
            },
          ],
          "id": 2,
          "name": "Leaf 2",
          "state": Object {
            "deletable": true,
            "expanded": true,
          },
        },
        Object {
          "id": 5,
          "name": "Leaf 5",
        },
      ],
      "id": 0,
      "name": "Leaf 1",
      "state": Object {
        "expanded": true,
      },
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": 7,
              "name": "Leaf 8",
            },
            Object {
              "id": 8,
              "name": "Leaf 9",
            },
          ],
          "id": 6,
          "name": "Leaf 7",
          "state": Object {
            "expanded": false,
          },
        },
        Object {
          "id": 9,
          "name": "Leaf 10",
        },
      ],
      "id": 1,
      "name": "Leaf 6",
      "state": Object {
        "deletable": true,
        "expanded": false,
      },
    },
    Object {
      "id": "z",
      "name": "Leaf z",
      "state": Object {
        "deletable": true,
        "favorite": true,
      },
    },
  ],
]
`;

exports[`TreeContainer extensions updateTypeHandlers should call injected prop onChange with the correct params for custom handlers 1`] = `
Array [
  Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "id": 3,
              "name": "Leaf 3",
              "state": Object {
                "deletable": true,
                "expanded": true,
                "favorite": true,
              },
            },
            Object {
              "children": Array [],
              "id": 4,
              "name": "Leaf 4",
              "state": Object {
                "expanded": true,
              },
            },
          ],
          "id": 2,
          "name": "Leaf 2",
          "state": Object {
            "deletable": true,
            "expanded": true,
          },
        },
        Object {
          "children": Array [],
          "id": 5,
          "name": "Leaf 5",
          "state": Object {
            "expanded": true,
          },
        },
      ],
      "id": 0,
      "name": "Leaf 1",
      "state": Object {
        "expanded": true,
      },
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "id": 7,
              "name": "Leaf 8",
              "state": Object {
                "expanded": true,
              },
            },
            Object {
              "children": Array [],
              "id": 8,
              "name": "Leaf 9",
              "state": Object {
                "expanded": true,
              },
            },
          ],
          "id": 6,
          "name": "Leaf 7",
          "state": Object {
            "expanded": true,
          },
        },
        Object {
          "children": Array [],
          "id": 9,
          "name": "Leaf 10",
          "state": Object {
            "expanded": true,
          },
        },
      ],
      "id": 1,
      "name": "Leaf 6",
      "state": Object {
        "deletable": true,
        "expanded": true,
      },
    },
    Object {
      "children": Array [],
      "id": "z",
      "name": "Leaf z",
      "state": Object {
        "deletable": true,
        "expanded": true,
        "favorite": true,
      },
    },
  ],
]
`;

exports[`TreeContainer extensions updateTypeHandlers should call injected prop onChange with the correct params for handler overrides 1`] = `
Array [
  Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": 3,
              "name": "Leaf 3",
              "state": Object {
                "deletable": true,
                "favorite": true,
              },
              "updated": true,
            },
            Object {
              "id": 4,
              "name": "Leaf 4",
            },
          ],
          "id": 2,
          "name": "Leaf 2",
          "state": Object {
            "deletable": true,
            "expanded": true,
          },
        },
        Object {
          "id": 5,
          "name": "Leaf 5",
        },
      ],
      "id": 0,
      "name": "Leaf 1",
      "state": Object {
        "expanded": true,
      },
    },
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "id": 7,
              "name": "Leaf 8",
            },
            Object {
              "id": 8,
              "name": "Leaf 9",
            },
          ],
          "id": 6,
          "name": "Leaf 7",
          "state": Object {
            "expanded": false,
          },
        },
        Object {
          "id": 9,
          "name": "Leaf 10",
        },
      ],
      "id": 1,
      "name": "Leaf 6",
      "state": Object {
        "deletable": true,
        "expanded": false,
      },
    },
    Object {
      "id": "z",
      "name": "Leaf z",
      "state": Object {
        "deletable": true,
        "favorite": true,
      },
    },
  ],
]
`;

exports[`TreeContainer should render a Tree with children and the correct props 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 3,
            "name": "Leaf 3",
            "state": Object {
              "deletable": true,
              "favorite": true,
            },
          },
          Object {
            "id": 4,
            "name": "Leaf 4",
          },
        ],
        "id": 2,
        "name": "Leaf 2",
        "state": Object {
          "deletable": true,
          "expanded": true,
        },
      },
      Object {
        "id": 5,
        "name": "Leaf 5",
      },
    ],
    "deepness": 0,
    "id": 0,
    "name": "Leaf 1",
    "parents": Array [],
    "state": Object {
      "expanded": true,
    },
  },
  Object {
    "children": Array [
      Object {
        "id": 3,
        "name": "Leaf 3",
        "state": Object {
          "deletable": true,
          "favorite": true,
        },
      },
      Object {
        "id": 4,
        "name": "Leaf 4",
      },
    ],
    "deepness": 1,
    "id": 2,
    "name": "Leaf 2",
    "parents": Array [
      0,
    ],
    "state": Object {
      "deletable": true,
      "expanded": true,
    },
  },
  Object {
    "deepness": 2,
    "id": 3,
    "name": "Leaf 3",
    "parents": Array [
      0,
      2,
    ],
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
  Object {
    "deepness": 2,
    "id": 4,
    "name": "Leaf 4",
    "parents": Array [
      0,
      2,
    ],
  },
  Object {
    "deepness": 1,
    "id": 5,
    "name": "Leaf 5",
    "parents": Array [
      0,
    ],
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "id": 7,
            "name": "Leaf 8",
          },
          Object {
            "id": 8,
            "name": "Leaf 9",
          },
        ],
        "id": 6,
        "name": "Leaf 7",
        "state": Object {
          "expanded": false,
        },
      },
      Object {
        "id": 9,
        "name": "Leaf 10",
      },
    ],
    "deepness": 0,
    "id": 1,
    "name": "Leaf 6",
    "parents": Array [],
    "state": Object {
      "deletable": true,
      "expanded": false,
    },
  },
  Object {
    "deepness": 0,
    "id": "z",
    "name": "Leaf z",
    "parents": Array [],
    "state": Object {
      "deletable": true,
      "favorite": true,
    },
  },
]
`;

exports[`TreeContainer when draggable should render a DraggingContainer 1`] = `
Object {
  "className": "tree-container",
  "onChange": [Function],
}
`;
